nginx-odoo.py

```diff
@@ -8,326 +8,172 @@
-    if not all(email_regex.match(t) for t in _to_list):
-        return False
-    if BRANDING:
-        msgtext = "{} security code: {}".format(BRANDING, code)
-    else:
-        msgtext = "Security code: {}".format(code)
-    msg = MIMEText(msgtext)
-    msg['Subject'] = msgtext
-    msg['From'] = SMTP_FROM
-    msg['To'] = _to
-    s = None
-    retries = 3
-    success = False
-    logging.info('Trying to send mail..')
-    while (not success) and retries > 0:
-        try:
-            s = smtp_connect()
-            smtp_login(s)
-            s.sendmail(SMTP_FROM, _to_list, msg.as_string())
-            s.quit()
-            s.close()
-            success = True
-            break
-        except smtplib.SMTPServerDisconnected:
-            retries -= 1
-    if not success:
-        logging.error('SMTP failed after three retries')
-        return False
-    logging.info('Mail with security code sent to %s', _to)
-    return True
-
-
-# Handle login
-@app.route('/', method='POST')
-def do_verify():
+    # Redirect to other service
+    # needed when using the bouncer to authenticate a user with the bouncer
+    redirect_url=self.get_argument('redirect',None)
+    if redirect_url!=None:
+      session_id=self.get_cookie('session_id')
+      if db.verify_session(session_id):
+        if not redirect_url.endswith('/'):
+          redirect_url+='/'
+        return self.redirect(f'{redirect_url}auth/{session_id}')
+    self.render(r'./templates/login.html',**config.theme_params)
+  def post(self):
     # handle username/password
     # TODO: CSRF protection
-    username = request.forms.get('username')
-    password = request.forms.get('password')
+    username=self.get_body_argument('username',default=None)
+    password=self.get_body_argument('password',default=None)
     if username and password:
-        logging.info('Verifying username %s and password...', username)
-        handler = OdooAuthHandler()
-        data, session_id = handler.check_login(username, password)
-        if session_id:
-            hotp = pyotp.HOTP(HOTP_SECRET)
-            counter, code = db.next_hotp_id(session_id)
-            key = hotp.at(counter)
-            if not send_mail(username, key):
-                message = 'Mail with security code not sent.'
-                logging.error(message)
-                return template('login', dict(theme_params, error=message))
-            return template(
-                'hotp', dict(theme_params.items(), counter=counter, code=code))
+      logging.info('Verifying username %s and password...', username)
+      handler = OdooAuthHandler()
+      data, session_id = handler.check_login(username, password)
+      if session_id:
+        hotp = pyotp.HOTP(config.HOTP_SECRET)
+        counter, code = db.next_hotp_id(session_id)
+        key = hotp.at(counter)
+        if config.disable_email:
+          # Display hotp in the log in stead of sending an email
+          logging.info(f'HOTP code: {key}')
         else:
```

Comment:
When I add here `import time; time.sleep(60)` to artificially slow down the sending of email, and I let the program send an email, then I cannot open a new login screen in a new tab until the mail is sent.

Thus, it seems we are still not opening the login screen async.

Reply:
this should work now. you can put time.sleep(60) in the email.send function, which is now async.
do we need any other functions to be async?

Reply:
Yes, OdooAuthHandler.check_login()